### PR TITLE
CLDC-2777: Set up SNS

### PR DESCRIPTION
**Pull request by SamSeed-Softwire**
Will do dev and prod once the rest of this PR is finalised.

### CODE REVIEW

File: terraform/modules/monitoring/policy.tf
Comment (SamSeed-Softwire) [inline terraform/modules/monitoring/policy.tf:1]: I could have also just used regular IAM policies here. See [here](https://docs.aws.amazon.com/sns/latest/dg/sns-using-identity-based-policies.html#iam-and-sns-policies) for a comparison of the approaches. In retrospect I think it might actually have been better to use regular IAM policies (rather than `aws_sns_topic_policy`) - like how it was done in EPB. The way I've done it here does allow for cross-account SNS publishing, but I don't think that will be needed.

I wanted to get a second opinion though about whether the other way was indeed better, and even if it was whether it was worth the time to change it (probably yes IMO).

File: terraform/modules/monitoring/policy.tf
Comment (SamSeed-Softwire) [inline terraform/modules/monitoring/policy.tf:27]: Currently nothing is allowed to publish - my plan was to add ARNs here for resources as they are created.

File: terraform/modules/monitoring/topic.tf
Comment (SamSeed-Softwire) [inline terraform/modules/monitoring/topic.tf:2]: AFAIK there will only be one SNS topic which we use for everything. Happy to use a disambiguating name now though if that's not (or might not) be the case.

File: terraform/modules/monitoring/policy.tf
Comment (RachaelBooth) [reply terraform/modules/monitoring/policy.tf:1]: Overall I don't think it matters - the main concern is probably how easy it is to add new things that can publish to the topic (particularly for e.g. review apps where we might end up with one topic and several instances of other things publishing notifications)

File: terraform/staging/main.tf
Comment (Chirag-Bhatti) [inline terraform/staging/main.tf:125]: Believe we might need to add this in dev and prod as well? 😊

Comment (SamSeed-Softwire) [reply terraform/staging/main.tf:125]: Ah, sorry, knew I'd forgotten to write something on the PR - basically I'm generally just doing stuff for staging, then once the PR has been approved and I know exactly what staging should look like only then do I replicate for dev and prod. Helps to reduce the amount of work I have to do!

File: terraform/modules/monitoring/topic.tf
Comment (Chirag-Bhatti) [reply terraform/modules/monitoring/topic.tf:2]: A single topic feels right to me (we're only using this for alarms / notifications to slack)

File: terraform/modules/monitoring/policy.tf
Comment (Chirag-Bhatti) [reply terraform/modules/monitoring/policy.tf:27]: Looks good (presuming it's the arn's of the `aws_cloudwatch_metric_alarm`'s that will be added here)

**Comment by Chirag-Bhatti** [comment]
@RachaelBooth / @SamSeed-Softwire, it might be out of scope for ticket/MVP, but I wanted to check if we should consider covering the case where the alarm can't be delivered to the subscriber from SNS? Should we setup a dead letter queue for example? Some info [here](https://docs.aws.amazon.com/sns/latest/dg/sns-dead-letter-queues.html)

I suppose the alarm not reaching Slack might not be so bad if we also send out email notifications for example. Also, if the alarm notification reaches much later because the issue preventing delivery was temporary, it might not be as useful anymore too. But we also may want to ensure an important alarm isn't missed because there was only one medium of knowing (outside of checking the AWS console)

**Comment by SamSeed-Softwire** [comment]
> @RachaelBooth / @SamSeed-Softwire, it might be out of scope for ticket/MVP, but I wanted to check if we should consider covering the case where the alarm can't be delivered to the subscriber from SNS? Should we setup a dead letter queue for example? Some info [here](https://docs.aws.amazon.com/sns/latest/dg/sns-dead-letter-queues.html)
> 
> I suppose the alarm not reaching Slack might not be so bad if we also send out email notifications for example. Also, if the alarm notification reaches much later because the issue preventing delivery was temporary, it might not be as useful anymore. But we may want to ensure an important alarm isn't missed, because there was only one medium of knowing (outside of checking the AWS console)
> 
> What's your thoughts?

@Chirag-Bhatti I think that sounds sensible. From reading about it it seems that DLQs are at the subscription-level, so even if we do want them they are out of scope for this ticket (which you might have already realised). If we do want DLQs then I guess we should do them as part of each ticket where we're creating a subscription.

File: terraform/modules/monitoring/policy.tf
Comment (RachaelBooth) [reply terraform/modules/monitoring/policy.tf:1]: Makes sense to me

**Review by RachaelBooth — APPROVED** [review]

**Comment by RachaelBooth** [comment]
> > @RachaelBooth / @SamSeed-Softwire, it might be out of scope for ticket/MVP, but I wanted to check if we should consider covering the case where the alarm can't be delivered to the subscriber from SNS? Should we setup a dead letter queue for example? Some info [here](https://docs.aws.amazon.com/sns/latest/dg/sns-dead-letter-queues.html)
> > I suppose the alarm not reaching Slack might not be so bad if we also send out email notifications for example. Also, if the alarm notification reaches much later because the issue preventing delivery was temporary, it might not be as useful anymore. But we may want to ensure an important alarm isn't missed, because there was only one medium of knowing (outside of checking the AWS console)
> > What's your thoughts?
> 
> @Chirag-Bhatti I think that sounds sensible. From reading about it it seems that DLQs are at the subscription-level, so even if we do want them they are out of scope for this ticket (which you might have already realised). If we do want DLQs then I guess we should do them as part of each ticket where we're creating a subscription.

The interesting thing about dead letter queues with this kind of use case is that our only subscription is going to be "notify us of alarms" and the way we'd find out that something had hit the dead letter queue would be a similar alarm and thus likely to be subject to the same problem. (Assuming it's not a message specific problem, but I wouldn't expect it to be for this type of message)

We ought to consider one at some point, perhaps with a different notification mechanism? I think it's relatively low down the priority list though.

**Pull request merged**
